middle_end/flambda/simplify/simplify_common.ml

```diff
@@ -50,6 +50,49 @@ type simplify_toplevel =
   -> exn_cont_scope:Scope.t
   -> Expr.t * Upwards_acc.t
 
+let is_self_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.closure_info denv with
+  | Not_in_a_closure -> false
+  | Closure { code_id = fun_code_id;
+              return_continuation = _; exn_continuation = _; } ->
+    begin match Apply.call_kind apply with
+    | Function (Direct { code_id = apply_code_id; _ }) ->
+      Code_id.equal fun_code_id apply_code_id
+    | Method _ | C_call _
+    | Function (Indirect_known_arity _ | Indirect_unknown_arity)
+      -> false
+    end
+  (* CR: This case should not happend, but it's relatively safe to
+     return false here *)
+  | In_a_set_of_closure_but_not_yet_in_a_specific_closure -> false
+
+let is_tail_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.closure_info denv with
+  | Not_in_a_closure -> false
+  | In_a_set_of_closure_but_not_yet_in_a_specific_closure ->
+    (* same as for {!is_self_call} *) false
+  | Closure { code_id = _;
```

Comment:
I'm not sure where these functions are being used, but could we get away with a single function that just detects self tail calls?  If so, that would be more efficient, as we wouldn't have to examine the closure_info and call_kind twice for example.

Reply:
These functions are not yet used anywhere, this PR is some groundwork for the future unrecursify work. In any case, since we don't have sets of recursive continuations, it's indeed not likely we'll try and unrecursify sets of mutually recursive functions any time soon, so the two separate functions are indeed not needed.
There is now a single `is_self_tail_cal` function.